File: clients/python/src/beebium/debugger.py
# Copyright 2025 Robert Smallshire <robert@smallshire.org.uk>
#
# This file is part of Beebium.
#
# Beebium is free software: you can redistribute it and/or modify it under the terms of the
# GNU General Public License as published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version. Beebium is distributed in the hope that it will
# be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
# You should have received a copy of the GNU General Public License along with Beebium.
# If not, see <https://www.gnu.org/licenses/>.

"""Debugger control interface for the beebium client."""

from __future__ import annotations

from dataclasses import dataclass

from beebium._proto import debugger_pb2, debugger_pb2_grpc
from beebium.exceptions import DebuggerError


@dataclass(frozen=True)
class ExecutionState:
    """Current execution state of the emulator."""

    is_running: bool
    cycle_count: int
    halt_reason: str
    sequence: int


@dataclass(frozen=True)
class Breakpoint:
    """A breakpoint set in the emulator."""

    id: int
    address: int


@dataclass(frozen=True)
class StepResult:
    """Result of a step operation."""

    success: bool
    error: str
    instructions_executed: int
    cycles_executed: int
    state: ExecutionState


class Debugger:
    """Debugger control interface.

    Provides execution control, stepping, and breakpoint management.

    Usage:
        # Stop execution
        state = bbc.debugger.stop()

        # Step through instructions
        result = bbc.debugger.step(10)

        # Set breakpoint and run until hit
        bp_id = bbc.debugger.add_breakpoint(0xC000)
        state = bbc.debugger.run_until(0xC000)
    """

    def __init__(self, stub: debugger_pb2_grpc.DebuggerControlStub):
        """Create a debugger interface.

        Args:
            stub: The gRPC stub for the DebuggerControl service.
        """
        self._stub = stub

    # Execution control

    def get_state(self) -> ExecutionState:
        """Get current execution state."""
        response = self._stub.GetState(debugger_pb2.Empty())
        return ExecutionState(
            is_running=response.is_running,
            cycle_count=response.cycle_count,
            halt_reason=response.halt_reason,
            sequence=response.sequence,
        )

    def run(self) -> None:
        """Resume execution.

        Raises:
            DebuggerError: If the operation fails.
        """
        response = self._stub.Run(debugger_pb2.Empty())
        if not response.success:
            raise DebuggerError(f"Failed to start execution: {response.error}")

    def stop(self) -> ExecutionState:
        """Pause execution.

        Returns:
            The execution state after stopping.
        """
        response = self._stub.Stop(debugger_pb2.Empty())
        if not response.success:
            raise DebuggerError("Failed to stop execution")
        return ExecutionState(
            is_running=response.state.is_running,
            cycle_count=response.state.cycle_count,
            halt_reason=response.state.halt_reason,
            sequence=response.state.sequence,
        )

    def reset(self) -> None:
        """Reset the machine.

        Raises:
            DebuggerError: If the reset fails.
        """
        response = self._stub.Reset(debugger_pb2.Empty())
        if not response.success:
            raise DebuggerError("Failed to reset machine")

    def step(self, count: int = 1) -> StepResult:
        """Step by instruction(s).

        Machine must be stopped first.

        Args:
            count: Number of instructions to execute.

        Returns:
            The step result including state after stepping.

        Raises:
            DebuggerError: If machine is running or step fails.
        """
        request = debugger_pb2.StepRequest(count=count)
        response = self._stub.StepInstruction(request)
        if not response.success:
            raise DebuggerError(f"Step failed: {response.error}")
        return StepResult(
            success=response.success,
            error=response.error,
            instructions_executed=response.instructions_executed,
            cycles_executed=response.cycles_executed,
            state=ExecutionState(
                is_running=response.state.is_running,
                cycle_count=response.state.cycle_count,
                halt_reason=response.state.halt_reason,
                sequence=response.state.sequence,
            ),
        )

    def step_cycles(self, count: int = 1) -> StepResult:
        """Step by CPU cycle(s).

        Machine must be stopped first.

        Args:
            count: Number of cycles to execute.

        Returns:
            The step result including state after stepping.

        Raises:
            DebuggerError: If machine is running or step fails.
        """
        request = debugger_pb2.StepRequest(count=count)
        response = self._stub.StepCycle(request)
        if not response.success:
            raise DebuggerError(f"Step cycles failed: {response.error}")
        return StepResult(
            success=response.success,
            error=response.error,
            instructions_executed=response.instructions_executed,
            cycles_executed=response.cycles_executed,
            state=ExecutionState(
                is_running=response.state.is_running,
                cycle_count=response.state.cycle_count,
                halt_reason=response.state.halt_reason,
                sequence=response.state.sequence,
            ),
        )

    # Convenience properties

    @property
    def is_running(self) -> bool:
        """True if the machine is currently running."""
        return self.get_state().is_running

    @property
    def is_stopped(self) -> bool:
        """True if the machine is currently stopped/paused."""
        return not self.is_running

    @property
    def cycle_count(self) -> int:
        """Current CPU cycle count."""
        return self.get_state().cycle_count

    # Breakpoints

    def add_breakpoint(self, address: int) -> int:
        """Add a breakpoint at the given address.

        Args:
            address: The address to break on (0x0000-0xFFFF).

        Returns:
            The breakpoint ID.

        Raises:
            DebuggerError: If the breakpoint cannot be added.
        """
        request = debugger_pb2.AddBreakpointRequest(address=address)
        response = self._stub.AddBreakpoint(request)
        if not response.success:
            raise DebuggerError(f"Failed to add breakpoint at ${address:04X}")
        return response.id

    def remove_breakpoint(self, breakpoint_id: int) -> bool:
        """Remove a breakpoint by ID.

        Args:
            breakpoint_id: The breakpoint ID returned by add_breakpoint().

        Returns:
            True if removed, False if not found.
        """
        request = debugger_pb2.RemoveBreakpointRequest(id=breakpoint_id)
        response = self._stub.RemoveBreakpoint(request)
        return response.success

    def list_breakpoints(self) -> list[Breakpoint]:
        """List all active breakpoints.

        Returns:
            List of active breakpoints.
        """
        response = self._stub.ListBreakpoints(debugger_pb2.Empty())
        return [
            Breakpoint(id=bp.id, address=bp.address) for bp in response.breakpoints
        ]

    def clear_breakpoints(self) -> int:
        """Remove all breakpoints.

        Returns:
            The number of breakpoints removed.
        """
        response = self._stub.ClearBreakpoints(debugger_pb2.Empty())
        return response.count_removed

    # Run-until helpers

    def run_until(
        self, address: int, timeout_cycles: int | None = None
    ) -> ExecutionState:
        """Run until the PC reaches the given address.

        Sets a temporary breakpoint, runs, waits for the breakpoint,
        and then clears the breakpoint.

        Args:
            address: The address to run until.
            timeout_cycles: Maximum cycles to run (not currently implemented).

        Returns:
            The execution state after hitting the address.

        Raises:
            DebuggerError: If the breakpoint cannot be set.
        """
        # Set temporary breakpoint
        bp_id = self.add_breakpoint(address)

        try:
            # Start execution
            self.run()

            # Wait for execution to stop (polling)
            import time

            while self.is_running:
                time.sleep(0.001)

            return self.get_state()
        finally:
            # Always remove the temporary breakpoint
            self.remove_breakpoint(bp_id)

File: clients/beebium-typescript-client/src/debugger.ts
/**
 * Debugger control interface for the Beebium TypeScript client.
 *
 * Provides execution control, stepping, and breakpoint management.
 */

import type { DebuggerControlClient } from "./generated/debugger.js";
import type {
    ExecutionState as ProtoExecutionState,
    ExecutionStateEvent as ProtoExecutionStateEvent,
    StepResponse as ProtoStepResponse,
    Breakpoint as ProtoBreakpoint,
} from "./generated/debugger.js";
import { StopReason } from "./generated/debugger.js";
import { promisify } from "./call-utils.js";
import { toAsyncIterable } from "./stream-utils.js";
import { DebuggerError, InvalidConditionError } from "./exceptions.js";

export interface ExecutionState {
    isRunning: boolean;
    cycleCount: number;
    haltReason: string;
    sequence: number;
}

export interface Breakpoint {
    id: number;
    startAddress: number;
    endAddress: number;
    condition: string;
    stopCounterpart: boolean;
    hitCount: number;
    enabled: boolean;
}

export interface ExecutionStateEvent {
    reason: StopReason;
    state: ExecutionState;
    message: string;
}

export type WatchpointType = "read" | "write" | "both";

export interface WatchpointInfo {
    id: number;
    startAddress: number;
    endAddress: number;
    type: WatchpointType;
    condition: string;
    stopCounterpart: boolean;
    hitCount: number;
    enabled: boolean;
}

export interface StepResult {
    success: boolean;
    error: string;
    instructionsExecuted: number;
    cyclesExecuted: number;
    state: ExecutionState;
}

export { StopReason };

function toExecutionState(proto: ProtoExecutionState): ExecutionState {
    return {
        isRunning: proto.isRunning,
        cycleCount: proto.cycleCount,
        haltReason: proto.haltReason,
        sequence: proto.sequence,
    };
}

function toStepResult(proto: ProtoStepResponse): StepResult {
    return {
        success: proto.success,
        error: proto.error,
        instructionsExecuted: proto.instructionsExecuted,
        cyclesExecuted: proto.cyclesExecuted,
        state: proto.state
            ? toExecutionState(proto.state)
            : { isRunning: false, cycleCount: 0, haltReason: "", sequence: 0 },
    };
}

function toBreakpoint(proto: ProtoBreakpoint): Breakpoint {
    return {
        id: proto.id,
        startAddress: proto.startAddress,
        endAddress: proto.endAddress,
        condition: proto.condition,
        stopCounterpart: proto.stopCounterpart,
        hitCount: proto.hitCount,
        enabled: proto.enabled,
    };
}

function toExecutionStateEvent(proto: ProtoExecutionStateEvent): ExecutionStateEvent {
    return {
        reason: proto.reason,
        state: proto.state
            ? toExecutionState(proto.state)
            : { isRunning: false, cycleCount: 0, haltReason: "", sequence: 0 },
        message: proto.message,
    };
}

/**
 * Debugger control interface.
 *
 * Provides execution control, stepping, and breakpoint management.
 */
export class Debugger {
    private readonly stub: DebuggerControlClient;

    constructor(stub: DebuggerControlClient) {
        this.stub = stub;
    }

    /** Get the current execution state. */
    async getState(): Promise<ExecutionState> {
        const response = await promisify<{}, ProtoExecutionState>(
            this.stub as unknown as Record<string, Function>,
            "getState",
            {},
        );
        return toExecutionState(response);
    }

    /** Start execution. Throws DebuggerError if the request is not successful. */
    async run(): Promise<void> {
        const response = await promisify<{}, { success: boolean; error: string }>(
            this.stub as unknown as Record<string, Function>,
            "run",
            {},
        );
        if (!response.success) {
            throw new DebuggerError(`run failed: ${response.error}`);
        }
    }

    /** Stop execution. Returns the execution state after stopping. */
    async stop(): Promise<ExecutionState> {
        const response = await promisify<{}, { success: boolean; state?: ProtoExecutionState }>(
            this.stub as unknown as Record<string, Function>,
            "stop",
            {},
        );
        if (!response.success) {
            throw new DebuggerError("stop failed");
        }
        if (!response.state) {
            throw new DebuggerError("stop returned no state");
        }
        return toExecutionState(response.state);
    }

    /** Reset the machine. */
    async reset(): Promise<void> {
        const response = await promisify<{}, { success: boolean }>(
            this.stub as unknown as Record<string, Function>,
            "reset",
            {},
        );
        if (!response.success) {
            throw new DebuggerError("reset failed");
        }
    }

    /** Step by instruction count. Returns the result of the step operation. */
    async step(count: number = 1): Promise<StepResult> {
        const response = await promisify<{ count: number }, ProtoStepResponse>(
            this.stub as unknown as Record<string, Function>,
            "stepInstruction",
            { count },
        );
        if (!response.success) {
            throw new DebuggerError(`step failed: ${response.error}`);
        }
        return toStepResult(response);
    }

    /** Step by cycle count. Returns the result of the step operation. */
    async stepCycles(count: number = 1): Promise<StepResult> {
        const response = await promisify<{ count: number }, ProtoStepResponse>(
            this.stub as unknown as Record<string, Function>,
            "stepCycle",
            { count },
        );
        if (!response.success) {
            throw new DebuggerError(`stepCycles failed: ${response.error}`);
        }
        return toStepResult(response);
    }

    /** Whether the machine is currently running. */
    async isRunning(): Promise<boolean> {
        const state = await this.getState();
        return state.isRunning;
    }

    /** Whether the machine is currently stopped. */
    async isStopped(): Promise<boolean> {
        const state = await this.getState();
        return !state.isRunning;
    }

    /**
     * Add a breakpoint on an address range. Returns the breakpoint ID.
     *
     * @param address - Start address (inclusive).
     * @param options.endAddress - End address (exclusive). 0 = address+1 (single address).
     *   Use 0x10000 for a full-range breakpoint that fires at every instruction boundary.
     * @param options.condition - Expression evaluated on hit. Empty = unconditional.
     * @param options.stopCounterpart - Signal the other processor to stop.
     */
    async addBreakpoint(
        address: number,
        options: { endAddress?: number; condition?: string; stopCounterpart?: boolean; enabled?: boolean } = {},
    ): Promise<number> {
        const { endAddress = 0, condition = "", stopCounterpart = false, enabled } = options;
        const request: Record<string, unknown> = { startAddress: address, endAddress, condition, stopCounterpart };
        if (enabled !== undefined) {
            request.enabled = enabled;
        }
        let response;
        try {
            response = await promisify<
                Record<string, unknown>,
                { success: boolean; id: number }
            >(
                this.stub as unknown as Record<string, Function>,
                "addBreakpoint",
                request,
            );
        } catch (e: unknown) {
            if (e instanceof Error && e.message.includes("INVALID_ARGUMENT")) {
                const detail = e.message.replace(/^.*INVALID_ARGUMENT:\s*/, "");
                throw new InvalidConditionError(detail || "Invalid condition expression");
            }
            throw e;
        }
        if (!response.success) {
            throw new DebuggerError(`addBreakpoint failed at address 0x${address.toString(16)}`);
        }
        return response.id;
    }

    /** Enable a breakpoint by ID. Throws DebuggerError if not found. */
    async enableBreakpoint(id: number): Promise<void> {
        const response = await promisify<{ id: number; enabled: boolean }, { success: boolean }>(
            this.stub as unknown as Record<string, Function>,
            "enableBreakpoint",
            { id, enabled: true },
        );
        if (!response.success) {
            throw new DebuggerError(`Breakpoint ${id} not found`);
        }
    }

    /** Disable a breakpoint by ID. Hit count and configuration are preserved. */
    async disableBreakpoint(id: number): Promise<void> {
        const response = await promisify<{ id: number; enabled: boolean }, { success: boolean }>(
            this.stub as unknown as Record<string, Function>,
            "enableBreakpoint",
            { id, enabled: false },
        );
        if (!response.success) {
            throw new DebuggerError(`Breakpoint ${id} not found`);
        }
    }

    /** Remove a breakpoint by ID. Returns true if the breakpoint was found and removed. */
    async removeBreakpoint(id: number): Promise<boolean> {
        const response = await promisify<{ id: number }, { success: boolean }>(
            this.stub as unknown as Record<string, Function>,
            "removeBreakpoint",
            { id },
        );
        return response.success;
    }

    /** List all current breakpoints. */
    async listBreakpoints(): Promise<Breakpoint[]> {
        const response = await promisify<{}, { breakpoints: ProtoBreakpoint[] }>(
            this.stub as unknown as Record<string, Function>,
            "listBreakpoints",
            {},
        );
        return response.breakpoints.map(toBreakpoint);
    }

    /** Clear all breakpoints. Returns the number of breakpoints removed. */
    async clearBreakpoints(): Promise<number> {
        const response = await promisify<{}, { countRemoved: number }>(
            this.stub as unknown as Record<string, Function>,
            "clearBreakpoints",
            {},
        );
        return response.countRemoved;
    }

    /**
     * Add a watchpoint on an address range. Returns the watchpoint ID.
     *
     * @param startAddress - Start address (inclusive).
     * @param endAddress - End address (exclusive).
     * @param type - Access type to watch.
     * @param options.condition - Expression evaluated on hit. Empty = unconditional.
     * @param options.stopCounterpart - Signal the other processor to stop.
     */
    async addWatchpoint(
        startAddress: number,
        endAddress: number,
        type: WatchpointType = "both",
        options: { condition?: string; stopCounterpart?: boolean; enabled?: boolean } = {},
    ): Promise<number> {
        const { condition = "", stopCounterpart = false, enabled } = options;
        const typeMap: Record<WatchpointType, number> = { read: 0, write: 1, both: 2 };
        const request: Record<string, unknown> = { startAddress, endAddress, type: typeMap[type], condition, stopCounterpart };
        if (enabled !== undefined) {
            request.enabled = enabled;
        }
        let response;
        try {
            response = await promisify<
                Record<string, unknown>,
                { success: boolean; id: number }
            >(
                this.stub as unknown as Record<string, Function>,
                "addWatchpoint",
                request,
            );
        } catch (e: unknown) {
            if (e instanceof Error && e.message.includes("INVALID_ARGUMENT")) {
                const detail = e.message.replace(/^.*INVALID_ARGUMENT:\s*/, "");
                throw new InvalidConditionError(detail || "Invalid condition expression");
            }
            throw e;
        }
        if (!response.success) {
            throw new DebuggerError(
                `addWatchpoint failed at 0x${startAddress.toString(16)}-0x${endAddress.toString(16)}`,
            );
        }
        return response.id;
    }

    /** Enable a watchpoint by ID. Throws DebuggerError if not found. */
    async enableWatchpoint(id: number): Promise<void> {
        const response = await promisify<{ id: number; enabled: boolean }, { success: boolean }>(
            this.stub as unknown as Record<string, Function>,
            "enableWatchpoint",
            { id, enabled: true },
        );
        if (!response.success) {
            throw new DebuggerError(`Watchpoint ${id} not found`);
        }
    }

    /** Disable a watchpoint by ID. Hit count and configuration are preserved. */
    async disableWatchpoint(id: number): Promise<void> {
        const response = await promisify<{ id: number; enabled: boolean }, { success: boolean }>(
            this.stub as unknown as Record<string, Function>,
            "enableWatchpoint",
            { id, enabled: false },
        );
        if (!response.success) {
            throw new DebuggerError(`Watchpoint ${id} not found`);
        }
    }

    /** Remove a watchpoint by ID. Returns true if found and removed. */
    async removeWatchpoint(id: number): Promise<boolean> {
        const response = await promisify<{ id: number }, { success: boolean }>(
            this.stub as unknown as Record<string, Function>,
            "removeWatchpoint",
            { id },
        );
        return response.success;
    }

    /** List all current watchpoints. */
    async listWatchpoints(): Promise<WatchpointInfo[]> {
        const typeMap: Record<number, WatchpointType> = { 0: "read", 1: "write", 2: "both" };
        const response = await promisify<
            {},
            { watchpoints: Array<{ id: number; startAddress: number; endAddress: number; type: number; condition: string; stopCounterpart: boolean; hitCount: number; enabled: boolean }> }
        >(
            this.stub as unknown as Record<string, Function>,
            "listWatchpoints",
            {},
        );
        return response.watchpoints.map((wp) => ({
            id: wp.id,
            startAddress: wp.startAddress,
            endAddress: wp.endAddress,
            type: typeMap[wp.type] ?? "both",
            condition: wp.condition ?? "",
            stopCounterpart: wp.stopCounterpart ?? false,
            hitCount: wp.hitCount ?? 0,
            enabled: wp.enabled ?? true,
        }));
    }

    /** Clear all watchpoints. Returns the number of watchpoints removed. */
    async clearWatchpoints(): Promise<number> {
        const response = await promisify<{}, { countRemoved: number }>(
            this.stub as unknown as Record<string, Function>,
            "clearWatchpoints",
            {},
        );
        return response.countRemoved;
    }

    /**
     * Subscribe to execution state change events as an async iterable.
     *
     * The server sends the current state immediately, then pushes events
     * whenever the execution state changes (breakpoint hit, manual stop,
     * run resumed, etc.).
     */
    async *watchExecutionState(): AsyncGenerator<ExecutionStateEvent> {
        const stream = this.stub.watchExecutionState({});
        for await (const event of toAsyncIterable(stream)) {
            yield toExecutionStateEvent(event);
        }
    }

    /**
     * Wait for the machine to stop executing.
     *
     * Subscribes to the execution state stream and waits for a transition
     * to the stopped state. If the machine is already stopped when this is
     * called, it waits until the machine runs and then stops again.
     *
     * @param timeoutMs - Maximum time to wait in milliseconds. Default 30000.
     */
    async waitForStop(timeoutMs = 30000): Promise<ExecutionStateEvent> {
        return new Promise((resolve, reject) => {
            const timer = setTimeout(() => {
                reject(new DebuggerError(`Timed out waiting for stop after ${timeoutMs}ms`));
            }, timeoutMs);

            (async () => {
                try {
                    let initialSequence: number | null = null;
                    for await (const event of this.watchExecutionState()) {
                        if (initialSequence === null) {
                            initialSequence = event.state.sequence;
                            if (event.state.isRunning) {
                                // Already running; wait for a stopped event.
                                continue;
                            }
                            // Initial state is stopped; wait for any
                            // subsequent stopped event (sequence will advance
                            // even if the running event is coalesced).
                            continue;
                        }
                        if (event.state.isRunning) {
                            continue;
                        }
                        // Stopped event with a later sequence than the initial
                        // snapshot means the machine ran and stopped.
                        if (event.state.sequence > initialSequence) {
                            clearTimeout(timer);
                            resolve(event);
                            return;
                        }
                    }
                    clearTimeout(timer);
                    reject(new DebuggerError("Stream ended without stop"));
                } catch (e) {
                    clearTimeout(timer);
                    reject(e);
                }
            })();
        });
    }

    /**
     * Run until execution reaches the given address.
     *
     * Sets a temporary breakpoint, subscribes to execution state events,
     * starts execution, and waits for the machine to stop.
     *
     * @param address - Address to break at.
     * @param timeoutMs - Maximum time to wait in milliseconds. Default 30000.
     */
    async runUntil(address: number, timeoutMs = 30000): Promise<ExecutionStateEvent> {
        const bpId = await this.addBreakpoint(address);
        try {
            // Ensure stopped, subscribe to the event stream, consume the
            // initial state, THEN start running. This avoids a race where
            // a tight-loop breakpoint fires before the stream is established.
            // Ensure stopped, subscribe, record sequence, then run.
            // Any stopped event with sequence > runSequence is our breakpoint.
            if (await this.isRunning()) await this.stop();
            const iter = this.watchExecutionState();
            const initial = await iter.next();
            const runSequence = initial.value!.state.sequence;
            await this.run();

            // Timeout: stop the machine to generate a STOP event that
            // wakes up the stream. We can't use iter.return() because
            // it can't interrupt an async generator blocked in await.
            let timedOut = false;
            const timer = setTimeout(() => {
                timedOut = true;
                this.stop().catch(() => {});
            }, timeoutMs);

            for await (const event of iter) {
                if (!event.state.isRunning && event.state.sequence > runSequence) {
                    clearTimeout(timer);
                    if (timedOut) {
                        throw new DebuggerError(`Timed out waiting for stop after ${timeoutMs}ms`);
                    }
                    return event;
                }
            }
            throw new DebuggerError(`Stream ended without stop`);
        } finally {
            await this.removeBreakpoint(bpId);
        }
    }
}
